### DISCUSSION TITLE
Translators in GA3H

### DISCUSSION THREAD

**Discussion by CaseySanchez** [laffernandes/gatl#29] [Q&A] [answered]
Hello,

I am trying to apply a translator to a point in GA3H as described in the following 3D Projective Geometric Algebra cheat sheet:
https://bivector.net/3DPGA.pdf

This example code appears to work using ganja.js (`!` is the dual operator, `>>>` is the sandwich operator):
```
// Translate the point at (1, 2, 3) by (1, 1, 1)

var point = !(1.0 * 1e0 + 1.0 * 1e1 + 2.0 * 1e2 + 3.0 * 1e3);
var translator = 1.0 + 1.0 * -0.5 * 1e01 + 1.0 * -0.5 * 1e02 + 1.0 * -0.5 * 1e03;

console.log("Point = " + (point));
console.log("Translated = " + (translator >>> point));
```

This outputs: 
```
Point = -3e_012+2e_013-e_023+e_123
Translated = -4e_012+3e_013-2e_023+e_123
```

The GATL equivalent yields puzzling results:
```
auto point = dual(1.0 * ep + 1.0 * e1 + 2.0 * e2 + 3.0 * e3);
auto translator = 1.0 + 1.0 * -0.5 * (ep ^ e1) + 1.0 * -0.5 * (ep ^ e2) + 1.0 * -0.5 * (ep ^ e3);
auto rotor = std::sqrt(2.0) * 0.5 * 0.5 + std::sqrt(2.0) * 0.5 * -0.5 * (e1 ^ e2);

std::cout << "Point = " << (point) << std::endl; 
std::cout << "Translated = " << (translator * point * inv(translator)) <<  std::endl;
std::cout << "Rotated = " << (rotor * point * inv(rotor)) << std::endl;
```

This outputs:
```
Point = (-1) * <e1^e2^e3> + 3 * <e1^e2^ep> + (-2) * <e1^e3^ep> + 1 * <e2^e3^ep>
Translated = 1.11022e-16 * <e1> + 0 * <e2> + (-5.55112e-17) * <e3> + 0 * <ep> + 3.28571 * <e1^e2^e3> + 1.85714 * <e1^e2^ep> + (-0.857143) * <e1^e3^ep> + (-0.142857) * <e2^e3^ep>
Rotated = 0 * <e3> + 0 * <ep> + (-1) * <e1^e2^e3> + 3 * <e1^e2^ep> + (-1) * <e1^e3^ep> + (-2) * <e2^e3^ep>
```

The behavior of the rotor is as expected, but I'm not certain as to what is going on with the translator.

Am I doing something incorrectly, or could this perhaps be a bug?

Thank you,
Casey

**Comment by laffernandes** [comment, marked as answer]
I believe the problem is the model of geometry used in both cases. I believe the problem is the model of geometry used in both cases.

Please check if with Ganja you are using PGA while with GATL you are using the homogeneous model of geometry. The difference between the two is in the metric of space. With PGA `e0 . e0 = 0` while in the homogenous model `e0 . e0 = 1`. As a consequence of the metric, the homogeneous model does not have translators, only rotors.

The inclusion of PGA in GATL is on my to-do list. For now, you can declate the model following the example available at https://github.com/laffernandes/gatl/blob/master/cpp/tools/example/miscelaneos/general_metric.cpp

For computational efficiency, replace expressions like `v * x * inv(v)` with `apply_even_versor(v, x)` or `apply_odd_versor(v, x)` or `apply_rotor(v, x)`.

**Reply by CaseySanchez** [reply]
Thank you for the information Dr. Fernandes. I mistakenly believed homogeneous and projective GAs to be one in the same.

**Reply by laffernandes** [reply]
I just included PGA in GATL. But I didn't test it yet. Use the `ga3p.h` header to access it.
